pkg/nvpci/nvpci.go

```diff
@@ -206,6 +234,16 @@ func (p *nvpci) GetAllDevices() ([]*NvidiaPCIDevice, error) {
 
 // GetGPUByPciBusID constructs an NvidiaPCIDevice for the specified address (PCI Bus ID).
 func (p *nvpci) GetGPUByPciBusID(address string) (*NvidiaPCIDevice, error) {
+	return p.getGPUByPciBusID(address, nil)
+}
```

Comment:
Why do you pass `nil` for the cache here? Doesn't that mean that none of the calls through the top-level API will utilize the cache? Instead of passing the cache through at all, can't we just hold it in the `nvpci` type and access it / update it as necessary?

Reply:
all funcs from this interface beside `GetGPUByPciBusID` uses `GetAllDevices` underneath:
```go
type Interface interface {
	GetAllDevices() ([]*NvidiaPCIDevice, error)
	Get3DControllers() ([]*NvidiaPCIDevice, error)
	GetVGAControllers() ([]*NvidiaPCIDevice, error)
	GetNVSwitches() ([]*NvidiaPCIDevice, error)
	GetGPUs() ([]*NvidiaPCIDevice, error)
	GetGPUByIndex(int) (*NvidiaPCIDevice, error)
	GetGPUByPciBusID(string) (*NvidiaPCIDevice, error)
	GetNetworkControllers() ([]*NvidiaPCIDevice, error)
	GetPciBridges() ([]*NvidiaPCIDevice, error)
	GetDPUs() ([]*NvidiaPCIDevice, error)
}
```
I didn't want to hold the cache inside the struct as we don't have any eventing system to know when something changes like `driver` of `numvfs`. The cache in here is just to store physical functions per each `GetAllDevices` call not to reread parent info for virtual functions created on the same physical function . In my testing on a machine with a lot of virtual functions, the implementation without cache takes almost two times longer.

Reply:
OK. Do we really need to pass the cache and do the lookup inside `GetGPUByPciBusID()` then? Can't the assignment / lookup in the cache be done completely inside `GetAllDevices()`, skipping calls to `GetGPUByPciBusID()` if they are already in the cache?

Reply:
That was my first idea. But we use recurrence for VFs and if given device has `physfn` directory we make another call to `getGPUByPciBusID()` inside  `getGPUByPciBusID`:
```go
		physFn, err := p.getGPUByPciBusID(filepath.Base(physFnAddress), cache)
```
to move the cache outside of `getGPUByPciBusID` we would have to get rid of this recurrence and return some ref to physical GPU(like only fill out `Address` of `ndevice.SriovInfo.VirtualFunction.PhysicalFuntion`)  and resolve it (from cache or disk) outside of `getGPUByPciBusID` e.g.:
```go
	cache := make(map[string]*NvidiaPCIDevice)
	for _, deviceDir := range deviceDirs {
		deviceAddress := deviceDir.Name()
		nvdevice, err := p.getGPUByPciBusID(deviceAddress)
		if err != nil {
			return nil, fmt.Errorf("error constructing NVIDIA PCI device %s: %v", deviceAddress, err)
		}
		if nvdevice == nil {
			continue
		}
		if ndevice.IsVf() {
		     physDeviceAddress := ndevice.SriovInfo.VirtualFunction.PhysicalFuntion.Address
		     physDevice, ok := cache[physDeviceAddress]
		     if !ok {
		          physDevice, err = p.getGPUByPciBusID(physDeviceAddress)
		          if err != nil {
		               return nil, err
		          }
		          cache[physDeviceAddress] = physDevice
		     }
		     ndevice.SriovInfo.VirtualFunction.PhysicalFunction = physDevice
		} else if ndevice.IsPf() {
		    cache[address] = ndevice
		}
		nvdevices = append(nvdevices, nvdevice)
	}

```

Reply:
Won't that break things for the case of calling `GetGPUByPciBusID()` outside of `GetAllDevices()`?

Reply:
OK; it's probably not worth arguing over anymore then. It's not terrible, its just not intuitive at first glance. Maybe add a comment as to why you pass `nil` for the cache in `GetGPUByPciBusID()` so as to force a relookup.